air/src/execution_step/boxed_value/stream/values_matrix.rs

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2023 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use air_interpreter_data::GenerationIdx;
+
+use typed_index_collections::TiVec;
+
+/// It is intended to store values according to generations which stream has. Each generation could
+/// contain several values, generation are ordered, meaning that values in elder generations are
+/// handled by AIR instructions (such as fold) later. Also, there is an order between values in one
+/// generation. And being placed by generations values could be considered as a matrix.
+///
+/// This matrix is used for values in previous and current data.
+#[derive(Debug, Clone)]
+pub(crate) struct ValuesMatrix<T> {
+    /// The first Vec represents generations, the second values in a generation. Generation is a set
+    /// of values that interpreter obtained from one particle. It means that number of generation on
+    /// a peer is equal to number of the interpreter runs in context of one particle.
+    values: TiVec<GenerationIdx, Vec<T>>,
+}
+
+impl<T> ValuesMatrix<T> {
+    pub fn new() -> Self {
+        Self { values: TiVec::new() }
+    }
+
+    pub fn remove_empty_generations(&mut self) {
+        self.values.retain(|generation| !generation.is_empty())
+    }
+
+    pub fn len(&self) -> GenerationIdx {
```

Comment:
The name looks misguiding. You ask for len() and return GenerationIdx.

Reply:
Do you have any ideas in mind?

Reply:
what about `generations_count`?

Reply:
I would call it last_generation but generation_count also works.